# Coursera-Genome-Assembly-Programming-Challenge
Assembled Phi-X174 genome using Overlap Graph, Kmer Composition and De-Bruijn Graph.

<img width="794" alt="default" src="https://user-images.githubusercontent.com/33269462/44305590-1b413b00-a349-11e8-8839-915bfda8b2af.png">


## phi X 174

The phi X 174 (or ΦX174) bacteriophage is a single-stranded DNA (ssDNA) virus and the first DNA-based genome to be sequenced. This work was completed by Fred Sanger and his team in 1977. In 1962, Walter Fiers and Robert Sinsheimer had already demonstrated the physical, covalently closed circularity of ΦX174 DNA. Nobel prize winner Arthur Kornberg used ΦX174 as a model to first prove that DNA synthesized in a test tube by purified enzymes could produce all the features of a natural virus, ushering in the age of synthetic biology. In 2003, it was reported by Craig Venter's group that the genome of ΦX174 was the first to be completely assembled in vitro from synthesized oligonucleotides. The ΦX174 virus particle has also been successfully assembled in vitro. Recently, it was shown how its highly overlapping genome can be fully decompressed and still remain functional.

<img width="259" alt="default" src="https://user-images.githubusercontent.com/33269462/44305606-65c2b780-a349-11e8-9ee5-ccb3e108aa3c.png">

## Problem Description:
### Input: 
A collection of Strings called reads of the original genome. Each read is a sub-string of the original genome(Genome can be circular also).

### Output:
A string S of minimum length that cantains all the strings(reads) given in the input as its sub-strings.
